### PR TITLE
feat: support Health Check in BackendTrafficPolicy

### PR DESCRIPTION
**Pull request by lemonlinger**
**What type of PR is this?**

**What this PR does / why we need it**:

This PR introduces active health checking to BackendTrafficPolicy, supporting 3 types of health checkers: HTTP, gRPC, and TCP.

**Which issue(s) this PR fixes**:

relates to https://github.com/envoyproxy/gateway/issues/2126

### CODE REVIEW

**Review by shawnh2 — COMMENTED** [review]
thanks @lemonlinger , it's better to add some e2e test and user facing doc in the following PR.

File: api/v1alpha1/healthcheck_types.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// HealthCheck defines the health check configuration.
+type HealthCheck struct {
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:11]: are all these values `required` ?
imo they should all be optional with sane defaults except for `healthChecker`

File: api/v1alpha1/healthcheck_types.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+	//
+	// +kubebuilder:validation:Minimum=1
+	UnhealthyThreshold uint32 `json:"unhealthyThreshold"`
+
+	// HealthyThreshold defines the number of healthy health checks required before a backend host is marked healthy.
+	//
+	// +kubebuilder:validation:Minimum=1
+	HealthyThreshold uint32 `json:"healthyThreshold"`
+
+	// HealthChecker defines the concrete health checker to do health checking.
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:32]: we need a better name here, it currently looks like
```
healthCheck:
  healthChecker:
  .....
```

File: api/v1alpha1/healthcheck_types.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+// HTTPHealthChecker defines the settings of http health check.
+type HTTPHealthChecker struct {
+	// Path defines the HTTP path that will be requested during health checking.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1024
+	Path string `json:"path" yaml:"path"`
+	// Method defines the HTTP method used for health checking.
+	// +optional
+	Method string `json:"method,omitempty" yaml:"method,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:83]: optional fields are always pointer type

File: api/v1alpha1/healthcheck_types.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	TCP *TCPHealthChecker `json:"tcp,omitempty" yaml:"tcp,omitempty"`
+}
+
+// HTTPHealthChecker defines the settings of http health check.
+type HTTPHealthChecker struct {
+	// Path defines the HTTP path that will be requested during health checking.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1024
+	Path string `json:"path" yaml:"path"`
+	// Method defines the HTTP method used for health checking.
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:81]: maybe add a note saying, defaults to `GET`

File: api/v1alpha1/healthcheck_types.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	// +optional
+	Method string `json:"method,omitempty" yaml:"method,omitempty"`
+	// ExpectedStatuses defines a list of HTTP response statuses considered healthy.
+	// +optional
+	ExpectedStatuses []Int64Range `json:"expectedStatuses,omitempty" yaml:"expectedStatuses,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:86]: range doesnt seem like the common case for me, prefer a list of `ints` []int{200, 201}` 
curious what others think

File: api/v1alpha1/healthcheck_types.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	// +optional
+	TCP *TCPHealthChecker `json:"tcp,omitempty" yaml:"tcp,omitempty"`
+}
+
+// HTTPHealthChecker defines the settings of http health check.
+type HTTPHealthChecker struct {
+	// Path defines the HTTP path that will be requested during health checking.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1024
+	Path string `json:"path" yaml:"path"`
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:80]: should it be `Path` or `Url` ?

Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:80]: how is the user providing the destination host name ?

File: api/v1alpha1/healthcheck_types.go [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+// +union
+//
+// +kubebuilder:validation:XValidation:rule="self.type == 'HTTP' ? has(self.http) : !has(self.http)",message="If Health Checker type is HTTP, http field needs to be set."
+// +kubebuilder:validation:XValidation:rule="self.type == 'GRPC' ? has(self.grpc) : !has(self.grpc)",message="If Health Checker type is GRPC, grpc field needs to be set."
+// +kubebuilder:validation:XValidation:rule="self.type == 'TCP' ? has(self.tcp) : !has(self.tcp)",message="If Health Checker type is TCP, tcp field needs to be set."
+type HealthChecker struct {
+	// Type defines the type of health checker.
+	// +kubebuilder:validation:Enum=HTTP;GRPC;TCP
+	// +unionDiscriminator
+	Type HealthCheckerType `json:"type" yaml:"type"`
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:60]: is there a lot of value in splitting up HTTP and GRPC ?
imo 2 types should be sufficient where response is checked for HTTP and conn establishment is checked for TCP

File: api/v1alpha1/healthcheck_types.go (hunk 6 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:80]: > how is the user providing the destination host name ?

Envoy will use the address of each endpoint as the host name. If you are referring to the Host header, envoy offers an optional Host field to set this. Should it be supported?

File: api/v1alpha1/healthcheck_types.go (hunk 2 above)
Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:32]: Thoughts on endpoint ?

File: api/v1alpha1/healthcheck_types.go (hunk 4 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:89]: Fixed

File: api/v1alpha1/healthcheck_types.go (hunk 3 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go]: Fixed

File: api/v1alpha1/healthcheck_types.go (hunk 2 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:40]: I feel a little bit confused.  Any thoughts from others?

File: api/v1alpha1/healthcheck_types.go (hunk 1 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:11]: Fixed.

File: internal/ir/xds.go [hunk 8]
```diff
@@ -1095,3 +1108,193 @@ type SlowStart struct {
...
+		matchCount++
+	}
+	if h.GRPC != nil {
+		matchCount++
+	}
+	if h.TCP != nil {
+		matchCount++
+	}
+	if matchCount != 1 {
+		errs = multierror.Append(errs)
```
[the remark this answers is not in the document]
Comment (lemonlinger) [reply internal/ir/xds.go]: Fixed

File: internal/xds/translator/cluster.go [hunk 9]
```diff
@@ -130,6 +133,104 @@ func buildXdsCluster(args *xdsClusterArgs) *clusterv3.Cluster {
...
+			grpcChecker.InitialMetadata = headerValues
+			hc.HealthChecker = &corev3.HealthCheck_GrpcHealthCheck_{
+				GrpcHealthCheck: grpcChecker,
+			}
+		}
+		if args.healthCheck.TCP != nil {
+			tcpChecker := &corev3.HealthCheck_TcpHealthCheck{}
+			if args.healthCheck.TCP.Send != nil {
+				var send corev3.HealthCheck_Payload
+				if args.healthCheck.TCP.Send.Text != "" {
```
[the remark this answers is not in the document]
Comment (lemonlinger) [reply internal/xds/translator/cluster.go]: Done

**Comment by lemonlinger** [comment]
> thanks @lemonlinger , it's better to add some e2e test and user facing doc in the following PR.

Sure, Will do.

File: api/v1alpha1/healthcheck_types.go [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+	HealthCheckerTypeGRPC HealthCheckerType = "GRPC"
+	// HealthCheckerTypeTCP defines the TCP type of health checking.
+	HealthCheckerTypeTCP HealthCheckerType = "TCP"
+)
+
+// HealthChecker defines the configuration of concrete health checker.
+// EG supports various types of health checking including HTTP, GRPC, TCP.
+// +union
+//
+// +kubebuilder:validation:XValidation:rule="self.type == 'HTTP' ? has(self.http) : !has(self.http)",message="If Health Checker type is HTTP, http field needs to be set."
```
[the remark this answers is not in the document]
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:61]: Done

File: test/cel-validation/backendtrafficpolicy_test.go [hunk 11]
```diff
@@ -16,6 +16,7 @@ import (
 	"time"
 
 	egv1a1 "github.com/envoyproxy/gateway/api/v1alpha1"
+	"github.com/envoyproxy/gateway/internal/utils/ptr"
```
[the remark this answers is not in the document]
Comment (lemonlinger) [reply test/cel-validation/backendtrafficpolicy_test.go]: Done

File: api/v1alpha1/healthcheck_types.go [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+	// +optional
+	Send *HealthCheckPayload `json:"send,omitempty" yaml:"send,omitempty"`
+	// Receive defines the expected response payloads.
+	// +optional
+	Receive []HealthCheckPayload `json:"receive,omitempty" yaml:"receive,omitempty"`
+}
+
+// Int64Range defines the int64 start and end of the range using half-open interval semantics [start, end).
+// +kubebuilder:validation:XValidation:rule="self.start <= self.end",message="start should be not greater than end"
+type Int64Range struct {
```
[the remark this answers is not in the document]
Comment (shawnh2) [reply api/v1alpha1/healthcheck_types.go:125]: sure, sounds good to me

[the remark this answers is not in the document]
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go]: Done

**Review by Xunzhuo — COMMENTED** [review]
Great work @lemonlinger, will start reviewing this week.

**Comment by Xunzhuo** [comment]
Hey @AliceProxy, would you like to take a review on this ?

File: api/v1alpha1/healthcheck_types.go (hunk 2 above)
Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:40]: how about removing `healthChecking` and copying the fields directly into top level struct ?

File: api/v1alpha1/healthcheck_types.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+	// Path defines the HTTP path that will be requested during health checking.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1024
+	Path string `json:"path" yaml:"path"`
+	// Method defines the HTTP method used for health checking.
+	// Defaults to GET
+	// +optional
+	Method *string `json:"method,omitempty" yaml:"method,omitempty"`
+	// ExpectedStatuses defines a list of HTTP response statuses considered healthy.
+	// +optional
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:94]: does this default to `200` and `201` ?

File: api/v1alpha1/healthcheck_types.go [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+	Send *HealthCheckPayload `json:"send,omitempty" yaml:"send,omitempty"`
+	// Receive defines the expected response payloads.
+	// +optional
+	Receive []HealthCheckPayload `json:"receive,omitempty" yaml:"receive,omitempty"`
+}
+
+// HTTPStatusRange defines the start and end of the http status range using half-open interval semantics [start, end). Only statuses in the range [100, 600) are allowed.
+// +kubebuilder:validation:XValidation:rule="self.start <= self.end",message="start should be not greater than end"
+type HTTPStatusRange struct {
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:125]: imo prefer a list of values instead of range

File: api/v1alpha1/healthcheck_types.go [hunk 15]
```diff
@@ -0,0 +1,163 @@
...
+	// Method defines the HTTP method used for health checking.
+	// Defaults to GET
+	// +optional
+	Method *string `json:"method,omitempty" yaml:"method,omitempty"`
+	// ExpectedStatuses defines a list of HTTP response statuses considered healthy.
+	// +optional
+	ExpectedStatuses []HTTPStatusRange `json:"expectedStatuses,omitempty" yaml:"expectedStatuses,omitempty"`
+	// ExpectedResponses defines a list of HTTP expected responses to match.
+	// +optional
+	ExpectedResponses []HealthCheckPayload `json:"expectedResponses,omitempty" yaml:"expectedResponses,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:98]: why is this a list ? is this a OR match ?

File: api/v1alpha1/healthcheck_types.go [hunk 16]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// TCPHealthChecker defines the settings of tcp health check.
+type TCPHealthChecker struct {
+	// Send defines the request payload.
+	// +optional
+	Send *HealthCheckPayload `json:"send,omitempty" yaml:"send,omitempty"`
+	// Receive defines the expected response payloads.
+	// +optional
+	Receive []HealthCheckPayload `json:"receive,omitempty" yaml:"receive,omitempty"`
```
Comment (arkodg) [inline api/v1alpha1/healthcheck_types.go:120]: same as https://github.com/envoyproxy/gateway/pull/2244/files#r1430780557

File: api/v1alpha1/healthcheck_types.go (hunk 6 above)
Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:88]: how can you reach a path without specifying the `host`

File: api/v1alpha1/healthcheck_types.go (hunk 7 above)
Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:68]: prefer if we only implement HTTP and TCP for now

File: api/v1alpha1/healthcheck_types.go (hunk 13 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:94]: default to 200.

File: api/v1alpha1/healthcheck_types.go (hunk 15 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:98]: According to [Envoy doc](https://www.envoyproxy.io/docs/envoy/v1.28.0/api-v3/config/core/v3/health_check.proto#envoy-v3-api-msg-config-core-v3-healthcheck-tcphealthcheck), it's a AND match. 

Perhaps in most cases, we only need one?

File: api/v1alpha1/healthcheck_types.go (hunk 14 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:125]: OK. I'll fix it later.

File: api/v1alpha1/healthcheck_types.go (hunk 15 above)
Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:98]: can you think of use cases where we'd need multiple matches ?

File: api/v1alpha1/healthcheck_types.go (hunk 14 above)
Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:125]: this is an API change, changes can be backwards compatible, we should pick the most common approach

File: api/v1alpha1/healthcheck_types.go (hunk 13 above)
Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:94]: can you add that into the doc string ?

File: api/v1alpha1/healthcheck_types.go (hunk 15 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:98]: not really.

Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:98]: lets make it singular then

Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go]: done

File: api/v1alpha1/healthcheck_types.go (hunk 14 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go]: done

File: api/v1alpha1/healthcheck_types.go (hunk 16 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go]: done

File: api/v1alpha1/healthcheck_types.go (hunk 13 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:95]: done

**Comment by arkodg** [comment]
@lemonlinger can you address the comments ?

**Comment by lemonlinger** [comment]
> @lemonlinger can you address the comments ?

Sure thing. I'll get it done today.

File: api/v1alpha1/healthcheck_types.go (hunk 2 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:40]: How about this ?

```yaml
  healthChecking: # optional
      connectionTimeout: duration # optional
      healthyCheckInterval: duration # optional, default=5s
      unhealthyCheckInterval: duration, # optional, default=10s
      unhealthyThreshold: int # optional, default=3
      healthyThreshold: int # optional, default=1
      type: Enum (grpc/http)
      grpc: # required when type=grpc
        upstreamName: string # required
        authority: string # optional
      http: # required when type=http
        expectedStatuses: # required, minItems=1
        - min: int # min=100, max=599
           max: int # min=100, max=599
        path: string # optional
        hostname: string # optional
```

Comment (arkodg) [reply api/v1alpha1/healthcheck_types.go:40]: yup looking good !
* prefer "healthCheck" as the top level noun
* lets only support tcp and http for now, and add support for grpc later

File: api/v1alpha1/healthcheck_types.go (hunk 7 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go:68]: OK

File: api/v1alpha1/healthcheck_types.go (hunk 2 above)
Comment (lemonlinger) [reply api/v1alpha1/healthcheck_types.go]: inline heathChecker to heathCheck and remove grpc checker.

done

**Review by arkodg — APPROVED** [review]
LGTM thanks !
can you also raise a separate GH issue for E2E and docs

**Review by Xunzhuo — APPROVED** [review]

**Comment by lemonlinger** [comment]
> LGTM thanks !
> 
> can you also raise a separate GH issue for E2E and docs

Sure thing.

**Pull request merged**